### DISCUSSION TITLE
OSX Sandbox - Container - Enable managed file access

### DISCUSSION THREAD

**Discussion by IcemarkUK** [axmolengine/axmol#1662] [Q&A] [answered]
In my Cocos2dx build for OSX I used to use the following function to get the folder to save data to..

```
 std::string getApplicationSupportPath()
{
   NSArray *paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
   NSString *documentsDirectory = [paths objectAtIndex:0];
   std::string strRet = [documentsDirectory UTF8String];
   strRet.append("/");
   return strRet;
}
```

The return value would be something like

`/Users/chris/Library/Containers/com.chilli.midnight/Data/Library/Application Support/`

In my CMake file I had the following
```
        set_target_properties(${APP_NAME} PROPERTIES
          XCODE_ATTRIBUTE_ENABLE_HARDENED_RUNTIME YES
          MACOSX_BUNDLE_INFO_PLIST "${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/mac/${APP_NAME}/Info.plist"
          XCODE_ATTRIBUTE_CODE_SIGN_ENTITLEMENTS "${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/mac/midnight.entitlements")
```

A combination of the Hardened runtime and entitlements turn on the sandbox which gives the containers folder.

I have the same in axmol project, but the above function now returns
` /Users/chris/Library/Application Support/com.chillihugger.midnight/`

When I build the project with Axmol build, I get this in the code signing
<img width="1130" alt="Screenshot 2024-02-09 at 20 33 43" src="https://github.com/axmolengine/axmol/assets/8301416/34f3da28-da17-43ad-ade7-8c13c92c35ea">

But with cocos2dx I get this
<img width="1130" alt="Screenshot 2024-02-09 at 20 27 11" src="https://github.com/axmolengine/axmol/assets/8301416/305f6229-f92f-4d49-acfe-249f22fd44bd">

If I change the options to match - it doesn't make any difference.

Any thoughts...?

**Comment by rh101** [comment]
> I have the same in axmol project, but the above function now returns
> ` /Users/chris/Library/Application Support/com.chillihugger.midnight/`

That path is the non-sandbox version, so it's not detecting that sandbox is enabled.

Do you have these lines in the `midnight.entitlements` file?

```
<key>com.apple.security.app-sandbox</key>
<true/>
```

Also, in your root `CMakeListst.txt`, do you have the following?

```
set_xcode_property(${APP_NAME} DEVELOPMENT_TEAM "XYZ1234567") # change to your dev team ID
set_xcode_property(${APP_NAME} CODE_SIGN_IDENTITY "") # leaving this blank should enable auto manage signing
```

It's most likely something in the `CMakeLists.txt` and/or the `proj.ios_mac/mac/Info.plist` files. Compare the differences between the cocos2d-x version and the axmol version.

**Reply by IcemarkUK** [reply]
I have the app-sandbox in my entitlements - it's exactly the same file that worked with Cocos2d-x

I had development team but not CODE_SIGN_IDENTITY - but I didn't have that in Cocos2d-x

Setting CODE_SIGN_IDENTITY does enable auto signing (Although I don't need it in my Cocos2d-x version) - so that bit at least worked - but sandbox is still not active.

I think it's somehow not picking up the entitlements file, but the setup of that's the same. I'l dig deeper and maybe look at the output of the build.

**Reply by rh101** [reply, marked as answer]
In `[axmol]/cmake/Modules/AXBuildHelpers.cmake`, take a look at the function `ax_setup_app_config`:

```
if(APPLE)
    # output macOS/iOS .app
    set_target_properties(${app_name} PROPERTIES MACOSX_BUNDLE 1)
    if(IOS AND (NOT ("${CMAKE_OSX_SYSROOT}" MATCHES ".*simulator.*")))
        set_xcode_property(${app_name} CODE_SIGNING_REQUIRED "YES")
        set_xcode_property(${app_name} CODE_SIGNING_ALLOWED "YES")
    else()
        # By default, explicit disable codesign for macOS PC
        set_xcode_property(${app_name} CODE_SIGN_IDENTITY "")
        set_xcode_property(${app_name} CODE_SIGNING_ALLOWED "NO") <= try commenting this line out, or set it to YES
        set_xcode_property(${app_name} CODE_SIGN_IDENTITY "NO") <= this line is strange? may not be correct
    endif()
elseif(WINDOWS)
```

Just out of curiosity, try commenting out all line `set_xcode_property(${app_name} CODE_SIGNING_ALLOWED "NO")`, and recreate the project files (`cmake build` etc.).  See if that helps with the sandbox issue.

When I get a chance to, I'll clone your repo and try it out on a Mac to see if I get the same output.

**Reply by IcemarkUK** [reply]
That didn't work - but thanks, it idea gives me something to check. I can compare axmol module cmake files against cocos versions.

**Reply by IcemarkUK** [reply]
Sorry, I didn't see the first comment on Signing Allowed.

So the problem is...

`set_xcode_property(${app_name} CODE_SIGNING_ALLOWED "NO")`

So adding the following in to my own CMakeList solves the problem.

```
set_xcode_property(${APP_NAME} CODE_SIGN_IDENTITY "")
set_xcode_property(${APP_NAME} CODE_SIGNING_ALLOWED "YES")
```

**Reply by rh101** [reply]
I'm glad it's all sorted. This info should be added to the FAQ at some point, because it's useful to know how to enable the app-sandbox for MacOS apps.
